ildd/ildd.1
.\" Copyright (c) 2002, 2003 Southern Storm Software, Pty Ltd.
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.TH ildd 1 "24 November 2003" "Southern Storm Software" "Portable.NET Development Tools"
.SH NAME
ildd \- print Intermediate Language (IL) program dependencies
.SH SYNOPSIS
.ll +8
.B ildd
[ options ]
.I input
\&...
.SH DESCRIPTION
.B Ildd
lists the assemblies and modules that are referenced by an IL object
file (.obj), an IL executable (.exe), or an IL dynamic link library (.dll).
.SH OPTIONS
.TP
.B \-f, \-\-file\-names
Print the name of each file that is processed, even if there is only
one file specified.
.TP
.B \-p, \-\-pinvoke
Print detailed information about PInvoke declarations.
.TP
.B \-r, \-\-recursive
Recursively print transitive dependencies.
.TP
.B \-\-help
Print a usage message for the \fBildd\fR program.
.TP
.B \-v, \-\-version
Print the version of the \fBildd\fR program.
.TP
.B \-\-
Marks the end of the command-line options, and the beginning of
the input filenames.  You may need to use this if your filename
begins with '-'.  e.g. "ildd -- -input.exe".  This is not needed
if the input is stdin: "ildd -" is perfectly valid.
.SH "AUTHOR"
Written by Southern Storm Software, Pty Ltd.

http://www.southern-storm.com.au/
.SH "SEE ALSO"
ilsize(1), ilfind(1), ilasm(1), cscc(1)
.SH "DIAGNOSTICS"
Exit status is 1 if an error occurred while processing the input.
Otherwise the exit status is 0.  See the manual page for
\fBilsize\fR(1) for a complete list of errors that may occur
when loading an IL image file.


NEWS

0.6.12 (16 January 2005)

	Runtime engine:

	* Beginning of PPC unroller implementation (Gopal V).

	C# Compiler:

	* Compatibility with /debug flags in csc (Gopal V).
	* Handling of method/property/etc declarations that hide base
	  class virtuals and access checks for same (Richard Baumann).

	Documentation:

	* Information on debugging the unroller with gdb (Gopal V).

	Platform Support:

	* Use a portable version of tparm to match Susv3 specs (Gopal V).
	* Case sensitivity during comparisons (Klaus Treichel).
	* EAX vs REG_EAX usage in interrupt_posix (Gopal V).
	* Fix cache flush logic for PPC (Gopal V).
	* Skip empty lines in "@file" response files (Stephen Compall).

0.6.10 (30 October 2004)

	Runtime engine:

	* Null parameter names when emitting parameters (Gopal V).
	* Various fixes to System.Reflection internalcalls (Gopal V).
	* Thread suspend/resume (Thong Nguyen).
	* Internalcalls for CustomAttributeBuilder (Rhys Weatherley).
	* Use "fprem", not "fprem1" for x86 IL remainder (Rhys Weatherley).
	* Handle vararg methods and delegates more centrally (Gopal V).
	* System.Decimal division bug (Klaus Treichel).
	* Do not inline loads from value types or crashes will occur (Gopal V).

	C# Compiler:

	* Work around some PPC64-related optimization problesm (Gopal V).
	* Nested class "famorassem" handling (Gopal V).
	* Fixes to enum initializers (Gopal V).

	BF Compiler:

	* Fix the behavior of BFOptimize() which currently destroys the
	  parse tree (Gopal V).

	Loader and Metadata:

	* Prevent double loading of assemblies by using the full path (Gopal V).
	* Use ILNativeUInt instead of "unsigned long" in il_types.h, which
	  fixes some problems with 64-bit platforms (Rhys Weatherley).
	* Properly import types before writing them out with SR.Emit (Gopal V).
	* Add a flag to enable ignoring load errors (Rhys Weatherley).

	Assembler:

	* Forward references for data labels (Gopal V).

	Platform Support:

	* Get user name via uid, rather than getlogin (Carl-Adam Brengesjo).
	* Enable threading for Solaris (Robin Kay).
	* Fix boot time calculation at the nanosecond level (Gopal V).
	* Bug in socket selection for infinite selects (Gopal V).
	* Alter gen_errno.sh to handle errno codes that are split across
	  multiple header files (Rhys Weatherley).
	* Response file support in cpp for systems with limited command
	  line lengths (Maciek Plewa).
	* Socket flags for Solaris (Adam J. Tybor).
	* Port socket code to hpux by including <sys/modem.h> (Gopal V).
	* Fix daylight savings for Win32 (Carl-Adam Brengesjo).
	* Libffi bug in x86-64, causing incorrect stack alignment (Gopal V).
	* Serial I/O routines for Win32 (Gopal V).
	* Timeouts in Unix version of serial reads (Adam Gibson).
	* Various profile fixes (Gopal V).
	* BeOS port (Tomas Wilhelmsson).
	* Cris port (Simon Posnjak).
	* Man page formatting problems (Peter Lund).

	Other:

	* Clean up Debian packaging (Russell Stuart).

0.6.8 (8 August 2004)

	Runtime engine:

	* Module.GetTypes and Assembly.GetExportedTypes internalcall (Gopal V).
	* Fixes to IL offset determination in the method cache (Gopal V).
	* "synchronized" method handling (Thong Nguyen).
	* Be careful not to notify finalizers during allocation (Thong Nguyen).
	* Inlining of Math functions (Thong Nguyen).
	* Inlining simple properties that get/set fields (Thong Nguyen).
	* Various fixes in the finalizer code (Thong Nguyen).
	* Thread aborting (Thong Nguyen).
	* Using signals to detect NULL references and the like (Thong Nguyen).
	* Thread execution contexts (Thong Nguyen).
	* Dns and Socket internalcalls (Gopal V).
	* GCAddrOfPinnedObject fixes (Llewellyn Pritchard).
	* Extended console support (Rhys Weatherley).

	C# Compiler:

	* Correct line numbers for argumentvar nodes (Gopal V).
	* Fixes to identifier resolution (Gopal V).
	* Coercion fixes in the arguments to user-defined operators (Gopal V).
	* Create nested classes in the proper order (Gopal V).
	* Null values and IntPtr (Gopal V).
	* Fixes for indexers on value types (Gopal V).
	* Assign add operations involving strings and objects (Gopal V).
	* Standalone assembly attributes (Rhys Weatherley).
	* Resolve classes before allocating heap objects (Gopal V).
	* Fix do ... while grammar to require a terminating ';' (Gopal V).
	* Add the "-fhidebyname" and "-fhidebysig" options (Rhys Weatherley).
	* Error message for private interface modifiers (Gopal V).
	* Check for event types before performing add/remove operations (Gopal V).

	C Compiler:

	* Fix ILNode_GetType for C strings (Gopal V).
	* Fix semantic analysis for member accesses (Gopal V).

	Loader and Metadata:

	* CompareUnicode not working correctly (Thong Nguyen).
	* ILClassImplements on a class should return true for itself (T. Nguyen).
	* Serialization of null strings in attribute blocks (Rhys Weatherley).
	* Recognise "OptionalAttribute" and "DefaultValueAttribute" (Rhys W).
	* Resolve classes in "TypeAssignCompatible" (Gopal V).
	* Recognise generic classes in files generated by .NET 2.0 (Rhys W).

	Linker:

	* Resource linking problems (Llewellyn Pritchard).
	* Obey "AssemblyVersionAttribute" if it is present (Rhys Weatherley).
	* "-fmetadata-version" option (Rhys Weatherley).
	* "-fculture" option (Rhys Weatherley).

	Platform Support:

	* Compilation fixes for 64-bit (Andrzej Dopierala).
	* Utilities for lists and queues (Thong Nguyen).
	* Various windows build fixes (Thong Nguyen).
	* ILGetSinceRebootTime for non-Windows platforms (Thong Nguyen).
	* Faster versions of volatile and interlocked functions (Thong Nguyen).
	* Daylight savings problem (David Logan).
	* Deal with trailing whitespace in response files (Rhys Weatherley).
	* OpenBSD support (Saul Hazledine).
	* Support for parallel builds (Gopal V).
	* Fix interlock functions for amd64 (Gopal V).

	Documentation:

	* Update the HACKING file (Klaus Treichel).
	* Typo in the ilgac manual (Richard Baumann).

0.6.6 (29 May 2004)

	Runtime engine:


ildd/ildd.c
/*
 * ildd.c - Print information about library dependencies.
 *
 * Copyright (C) 2002, 2003  Southern Storm Software, Pty Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <stdio.h>
#include "il_system.h"
#include "il_image.h"
#include "il_program.h"
#include "il_utils.h"

#ifdef	__cplusplus
extern	"C" {
#endif

/*
 * Table of command-line options.
 */
static ILCmdLineOption const options[] = {
	{"-f", 'f', 0, 0, 0},
	{"--file-names", 'f', 0,
		"--file-names or -f",
		"Print the name of each file that is processed."},
	{"-p", 'p', 0, 0, 0},
	{"--pinvoke", 'p', 0,
		"--pinvoke    or -p",
		"Print detailed information on PInvoke declarations."},
	{"-v", 'v', 0, 0, 0},
	{"--version", 'v', 0,
		"--version    or -v",
		"Print the version of the program."},
	{"--help", 'h', 0,
		"--help",
		"Print this help message."},
	{0, 0, 0, 0, 0}
};

static void usage(const char *progname);
static void version(void);
static int printDependencies(const char *filename, ILContext *context,
							 int multiple, int pinvoke);

int main(int argc, char *argv[])
{
	char *progname = argv[0];
	int filenames = 0;
	int pinvoke = 0;
	int sawStdin;
	int state, opt;
	char *param;
	int errors;
	int multiple;
	ILContext *context;

	/* Parse the command-line arguments */
	state = 0;
	while((opt = ILCmdLineNextOption(&argc, &argv, &state,
									 options, &param)) != 0)
	{
		switch(opt)
		{
			case 'f':
			{
				filenames = 1;
			}
			break;

			case 'p':
			{
				pinvoke = 1;
			}
			break;

			case 'v':
			{
				version();
				return 0;
			}
			/* Not reached */

			default:
			{
				usage(progname);
				return 1;
			}
			/* Not reached */
		}
	}

	/* We need at least one input file argument */
	if(argc <= 1)
	{
		usage(progname);
		return 1;
	}

	/* Create a context to use for image loading */
	context = ILContextCreate();
	if(!context)
	{
		fprintf(stderr, "%s: out of memory\n", progname);
		return 1;
	}

	/* Load and print information about the input files */
	sawStdin = 0;
	errors = 0;
	multiple = ((argc > 2) || filenames);
	while(argc > 1)
	{
		if(!strcmp(argv[1], "-"))
		{
			/* Dump the contents of stdin, but only once */
			if(!sawStdin)
			{
				errors |= printDependencies("-", context, multiple, pinvoke);
				sawStdin = 1;
			}
		}
		else
		{
			/* Dump the contents of a regular file */
			errors |= printDependencies(argv[1], context, multiple, pinvoke);
		}
		++argv;
		--argc;
	}

	/* Destroy the context */
	ILContextDestroy(context);
	
	/* Done */
	return errors;
}

static void usage(const char *progname)
{
	fprintf(stdout, "ILDD " VERSION " - IL Library Dependencies Utility\n");
	fprintf(stdout, "Copyright (c) 2002, 2003 Southern Storm Software, Pty Ltd.\n");
	fprintf(stdout, "\n");
	fprintf(stdout, "Usage: %s [options] input ...\n", progname);
	fprintf(stdout, "\n");
	ILCmdLineHelp(options);
}

static void version(void)
{

	fprintf(stdout, "ILDD " VERSION " - IL Library Dependencies Utility\n");
	fprintf(stdout, "Copyright (c) 2002, 2003 Southern Storm Software, Pty Ltd.\n");
	printf("\n");
	printf("ILDD comes with ABSOLUTELY NO WARRANTY.  This is free software,\n");
	printf("and you are welcome to redistribute it under the terms of the\n");
	printf("GNU General Public License.  See the file COPYING for further details.\n");
	printf("\n");
	printf("Use the `--help' option to get help on the command-line options.\n");
}

/*
 * Load an IL image from an input stream and print its dependency information.
 */
static int printDependencies(const char *filename, ILContext *context,
							 int multiple, int pinvoke)
{
	ILImage *image;
	ILAssembly *assem;
	ILModule *module;
	const char *name;
	const ILUInt16 *version;
	char *path;
	ILPInvoke *pinv;
	ILMethod *method;
	ILFileDecl *file;
	ILManifestRes *res;

	/* Attempt to load the image into memory */
	if(ILImageLoadFromFile(filename, context, &image,
					       IL_LOADFLAG_FORCE_32BIT |
						   IL_LOADFLAG_NO_RESOLVE, 1) != 0)
	{
		return 1;
	}

	/* Print the file header if we have multiple files */
	if(multiple)
	{
		if(!strcmp(filename, "-"))
		{
			fputs("stdin:\n", stdout);
		}
		else
		{
			fputs(filename, stdout);
			fputs(":\n", stdout);
		}
	}

	/* Print the assemblies that this file depends upon */
	assem = 0;
	while((assem = (ILAssembly *)ILImageNextToken
				(image, IL_META_TOKEN_ASSEMBLY_REF, assem)) != 0)
	{
		/* Print the assembly name and version */
		putc('\t', stdout);
		name = ILAssembly_Name(assem);
		version = ILAssemblyGetVersion(assem);
		fputs(name, stdout);
		printf("/%d:%d:%d:%d => ",
			   (int)(version[0]), (int)(version[1]),
			   (int)(version[2]), (int)(version[3]));

		/* Search for the full pathname of the referenced assembly */
		path = ILImageSearchPath(name, version, filename, 0, 0, 0, 0, 0, 0);
		if(path)
		{
			fputs(path, stdout);
			ILFree(path);
		}
		else
		{
			fputs("??", stdout);
		}
		putc('\n', stdout);
	}

	/* Print the module references that this file depends upon */
	module = 0;
	while((module = (ILModule *)ILImageNextToken
				(image, IL_META_TOKEN_MODULE_REF, module)) != 0)
	{
		printf("\tmodule %s\n", ILModule_Name(module));
	}

	/* Print the external file references that this file depends upon */
	file = 0;
	while((file = (ILFileDecl *)ILImageNextToken
				(image, IL_META_TOKEN_FILE, file)) != 0)
	{
		if(ILFileDecl_HasMetaData(file))
		{
			printf("\tfile %s => ", ILFileDecl_Name(file));
			path = ILImageSearchPath(ILFileDecl_Name(file), 0,
									 filename, 0, 0, 0, 0, 0, 0);
			if(path)
			{
				fputs(path, stdout);
				ILFree(path);
			}
			else
			{
				fputs("??", stdout);
			}
			putc('\n', stdout);
		}
		else
		{
			printf("\tfile %s\n", ILFileDecl_Name(file));
		}
	}

	/* Print the manifest resource information for the file */
	res = 0;
	while((res = (ILManifestRes *)ILImageNextToken
				(image, IL_META_TOKEN_MANIFEST_RESOURCE, res)) != 0)
	{
		if(ILManifestRes_OwnerFile(res) != 0)
		{
			printf("\tresource %s => file %s\n",
				   ILManifestRes_Name(res),
				   ILFileDecl_Name(ILManifestRes_OwnerFile(res)));
		}
		else if(ILManifestRes_OwnerAssembly(res) != 0)
		{
			printf("\tresource %s => assembly %s\n",
				   ILManifestRes_Name(res),
				   ILAssembly_Name(ILManifestRes_OwnerAssembly(res)));
		}
		else
		{
			printf("\tresource %s => (internal)\n", ILManifestRes_Name(res));
		}
	}

	/* Print the external PInvoke'd functions that this file depends upon */
	pinv = 0;
	while(pinvoke && (pinv = (ILPInvoke *)ILImageNextToken
				(image, IL_META_TOKEN_IMPL_MAP, pinv)) != 0)
	{
		/* Print the name of the PInvoke method */
		putc('\t', stdout);
		name = ILPInvoke_Alias(pinv);
		if(!name)
		{
			method = ILPInvokeGetMethod(pinv);
			if(method)
			{
				name = ILMethod_Name(method);
			}
			else
			{
				name = "??";
			}
		}
		fputs(name, stdout);
		fputs(" => ", stdout);

		/* Print the module that contains the PInvoke declaration */
		path = ILPInvokeResolveModule(pinv);
		if(path)
		{
			fputs(path, stdout);
			ILFree(path);
		}
		else
		{
			fputs("??", stdout);
		}
		putc('\n', stdout);
	}

	/* Clean up and exit */
	ILImageDestroy(image);
	return 0;
}

#ifdef	__cplusplus
};
#endif
